File: app/src/main/java/com/elementary/tasks/core/utils/TelephonyUtil.java
package com.elementary.tasks.core.utils;

import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.text.TextUtils;

import java.io.File;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public final class TelephonyUtil {

    private TelephonyUtil() {
    }

    public static void sendNote(@NonNull File file, Context context, @Nullable String message) {
        Intent intent = new Intent(Intent.ACTION_SEND);
        intent.setType("text/plain");
        String title = "Note";
        String note = "";
        if (message != null) {
            if (message.length() > 100) {
                title = message.substring(0, 48);
                title = title + "...";
            }
            if (message.length() > 150) {
                note = message.substring(0, 135);
                note = note + "...";
            }
        }
        intent.putExtra(Intent.EXTRA_SUBJECT, title);
        intent.putExtra(Intent.EXTRA_TEXT, note);
        Uri uri = UriUtil.getUri(context, file);
        intent.putExtra(Intent.EXTRA_STREAM, uri);
        intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        context.startActivity(Intent.createChooser(intent, "Send email..."));
    }

    public static void sendMail(Context context, @NonNull String email, String subject,
                                String message, String filePath) {
        Intent intent = new Intent(Intent.ACTION_SEND);
        intent.setType("text/plain");
        intent.putExtra(Intent.EXTRA_EMAIL, new String[]{email});
        intent.putExtra(Intent.EXTRA_SUBJECT, subject);
        intent.putExtra(Intent.EXTRA_TEXT, message);
        if (filePath != null) {
            Uri uri = UriUtil.getUri(context, filePath);
            intent.putExtra(Intent.EXTRA_STREAM, uri);
            intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        }
        context.startActivity(Intent.createChooser(intent, "Send email..."));
    }

    public static void sendSms(String number, Context context) {
        if (TextUtils.isEmpty(number)) {
            return;
        }
        Intent smsIntent = new Intent(Intent.ACTION_VIEW);
        smsIntent.setData(Uri.parse("sms:" + number));
        context.startActivity(smsIntent);
    }

    @SuppressWarnings("MissingPermission")
    public static void makeCall(String number, Context context) {
        if (TextUtils.isEmpty(number)) {
            return;
        }
        Intent callIntent = new Intent(Intent.ACTION_CALL);
        callIntent.setData(Uri.parse("tel:" + number));
        context.startActivity(callIntent);
    }

    public static void openApp(@NonNull String appPackage, Context context) {
        Intent launchIntent = context.getPackageManager().getLaunchIntentForPackage(appPackage);
        try {
            context.startActivity(launchIntent);
        } catch (ActivityNotFoundException ignored) {
        }
    }

    public static void openLink(@NonNull String link, Context context) {
        Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(link));
        try {
            context.startActivity(browserIntent);
        } catch (ActivityNotFoundException ignored) {
        }
    }

    public static void skypeCall(String number, Context context) {
        String uri = "skype:" + number + "?call";
        Intent sky = new Intent("android.intent.action.VIEW");
        sky.setData(Uri.parse(uri));
        context.startActivity(sky);
    }

    public static void skypeVideoCall(String number, Context context) {
        String uri = "skype:" + number + "?call&video=true";
        Intent sky = new Intent("android.intent.action.VIEW");
        sky.setData(Uri.parse(uri));
        context.startActivity(sky);
    }

    public static void skypeChat(String number, Context context) {
        String uri = "skype:" + number + "?chat";
        Intent sky = new Intent("android.intent.action.VIEW");
        sky.setData(Uri.parse(uri));
        context.startActivity(sky);
    }
}

File: app/src/main/java/com/elementary/tasks/core/utils/TelephonyUtil.kt
package com.elementary.tasks.core.utils

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.text.TextUtils
import android.widget.Toast
import com.elementary.tasks.BuildConfig
import com.elementary.tasks.R
import com.elementary.tasks.core.data.ui.UiShareData
import com.github.naz013.common.uri.UriUtil
import java.io.File

object TelephonyUtil {

  fun sendNote(file: File, context: Context, message: String?) {
    val intent = Intent(Intent.ACTION_SEND)
    intent.type = "text/plain"
    var title = "Note"
    var note = ""
    if (message != null) {
      if (message.length > 100) {
        title = message.take(48)
        title = "$title..."
      }
      if (message.length > 150) {
        note = message.take(135)
        note = "$note..."
      }
    }
    intent.putExtra(Intent.EXTRA_SUBJECT, title)
    intent.putExtra(Intent.EXTRA_TEXT, note)
    val uri = UriUtil.getUri(context, file, BuildConfig.APPLICATION_ID)
    intent.putExtra(Intent.EXTRA_STREAM, uri)
    intent.flags = Intent.FLAG_GRANT_READ_URI_PERMISSION
    try {
      context.startActivity(Intent.createChooser(intent, context.getString(R.string.share_send_email)))
    } catch (e: Exception) {
      Toast.makeText(context, R.string.app_not_found, Toast.LENGTH_SHORT).show()
    }
  }

  fun sendFile(file: File, context: Context, message: String?) {
    val intent = Intent(Intent.ACTION_SEND)
    intent.type = "*/*"
    intent.putExtra(Intent.EXTRA_SUBJECT, message)
    val uri = UriUtil.getUri(context, file, BuildConfig.APPLICATION_ID)
    intent.putExtra(Intent.EXTRA_STREAM, uri)
    intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
    try {
      val chooser = Intent.createChooser(intent, context.getString(R.string.share_send_email))
      context.startActivity(chooser)
    } catch (e: Exception) {
      Toast.makeText(context, R.string.app_not_found, Toast.LENGTH_SHORT).show()
    }
  }

  fun sendFile(context: Context, shareData: UiShareData) {
    if (shareData.file == null) return
    val intent = Intent(Intent.ACTION_SEND)
    intent.type = "*/*"
    intent.putExtra(Intent.EXTRA_SUBJECT, shareData.name)
    val uri = UriUtil.getUri(context, shareData.file, BuildConfig.APPLICATION_ID)
    intent.putExtra(Intent.EXTRA_STREAM, uri)
    intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
    try {
      val chooser = Intent.createChooser(intent, context.getString(R.string.share_send_email))
      context.startActivity(chooser)
    } catch (e: Exception) {
      Toast.makeText(context, R.string.app_not_found, Toast.LENGTH_SHORT).show()
    }
  }

  fun sendMail(
    context: Context,
    email: String,
    subject: String,
    message: String,
    filePath: String?
  ) {
    val intent = Intent(Intent.ACTION_SEND)
    intent.type = "text/plain"
    intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
    intent.putExtra(Intent.EXTRA_EMAIL, arrayOf(email))
    intent.putExtra(Intent.EXTRA_SUBJECT, subject)
    intent.putExtra(Intent.EXTRA_TEXT, message)
    if (filePath != null) {
      val uri = UriUtil.getUri(context, filePath, BuildConfig.APPLICATION_ID)
      intent.putExtra(Intent.EXTRA_STREAM, uri)
      intent.flags = Intent.FLAG_GRANT_READ_URI_PERMISSION
    }
    try {
      context.startActivity(Intent.createChooser(intent, context.getString(R.string.share_send_email)))
    } catch (e: Exception) {
      Toast.makeText(context, R.string.app_not_found, Toast.LENGTH_SHORT).show()
    }
  }

  fun sendMail(
    context: Context,
    email: String,
    subject: String,
    message: String,
    file: File?
  ) {
    val intent = Intent(Intent.ACTION_SEND)
    intent.type = "text/plain"
    intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
    intent.putExtra(Intent.EXTRA_EMAIL, arrayOf(email))
    intent.putExtra(Intent.EXTRA_SUBJECT, subject)
    intent.putExtra(Intent.EXTRA_TEXT, message)
    if (file != null) {
      val uri = UriUtil.getUri(context, file, BuildConfig.APPLICATION_ID)
      intent.putExtra(Intent.EXTRA_STREAM, uri)
      intent.flags = Intent.FLAG_GRANT_READ_URI_PERMISSION
    }
    try {
      context.startActivity(Intent.createChooser(intent, context.getString(R.string.share_send_email)))
    } catch (e: Exception) {
      Toast.makeText(context, R.string.app_not_found, Toast.LENGTH_SHORT).show()
    }
  }

  fun sendSms(number: String, context: Context) {
    if (TextUtils.isEmpty(number)) {
      return
    }
    val smsIntent = Intent(Intent.ACTION_VIEW)
    smsIntent.data = Uri.parse("sms:$number")
    smsIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
    try {
      context.startActivity(smsIntent)
    } catch (e: Exception) {
      Toast.makeText(context, R.string.app_not_found, Toast.LENGTH_SHORT).show()
    }
  }

  fun sendSms(context: Context, number: String, message: String?) {
    if (TextUtils.isEmpty(number)) {
      return
    }
    val smsIntent = Intent(Intent.ACTION_VIEW)
    smsIntent.data = Uri.parse("sms:$number")
    smsIntent.putExtra("sms_body", message)
    smsIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
    try {
      context.startActivity(smsIntent)
    } catch (e: Exception) {
      Toast.makeText(context, R.string.app_not_found, Toast.LENGTH_SHORT).show()
    }
  }

  fun makeCall(number: String, context: Context) {
    if (TextUtils.isEmpty(number)) {
      return
    }
    val callIntent = Intent(Intent.ACTION_CALL)
    callIntent.data = Uri.parse("tel:$number")
    callIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
    try {
      context.startActivity(callIntent)
    } catch (e: Exception) {
      Toast.makeText(context, R.string.app_not_found, Toast.LENGTH_SHORT).show()
    }
  }

  fun openApp(appPackage: String, context: Context) {
    val launchIntent = context.packageManager.getLaunchIntentForPackage(appPackage)
    try {
      context.startActivity(launchIntent)
    } catch (ignored: Exception) {
      Toast.makeText(context, R.string.app_not_found, Toast.LENGTH_SHORT).show()
    }
  }

  fun openLink(link: String, context: Context) {
    val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(link))
    try {
      context.startActivity(browserIntent)
    } catch (ignored: Exception) {
      Toast.makeText(context, R.string.app_not_found, Toast.LENGTH_SHORT).show()
    }
  }

  fun isPhoneNumber(target: String): Boolean {
    val phonePattern = "^[+]?[0-9 ()-]{3,25}\$".toRegex()
    return phonePattern.matches(target)
  }
}
